### PR TITLE
SONAR-7924 Modify WS api/qualitygates/select to accept project key

### PR DESCRIPTION
**Pull request by stas-vilchik**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/component/ComponentFinder.java [hunk 1]
```diff
@@ -56,6 +56,17 @@ public ComponentDto getByUuidOrKey(DbSession dbSession, @Nullable String compone
...
   }
 
+  public ComponentDto getByIdOrKey(DbSession dbSession, @Nullable Long componentId, @Nullable String componentKey, ParamNames parameterNames) {
+    checkArgument(componentId != null ^ componentKey != null, MSG_COMPONENT_ID_OR_KEY_TEMPLATE, parameterNames.getUuidParam(), parameterNames.getKeyParam());
+
+    if (componentId != null) {
+      return getById(dbSession, componentId);
+    }
+
+    checkArgument(!componentKey.isEmpty(), "The '%s' parameter must not be empty", parameterNames.getKeyParam());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ComponentFinder.java]: MAJOR NullPointerException might be thrown as 'componentKey' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [ComponentFinder.java#L51](https://github.com/SonarSource/sonarqube/blob/12f688d167828f4d46ce4445ab882d5e0b86fdb8/server/sonar-server/src/main/java/org/sonar/server/component/ComponentFinder.java#L51): Define a constant instead of duplicating this literal "The '%s' parameter must not be empty" 3 times. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1192)

File: server/sonar-server/src/main/java/org/sonar/server/component/ComponentFinder.java (hunk 1 above)
Comment (m-g-sonar) [inline server/sonar-server/src/main/java/org/sonar/server/component/ComponentFinder.java]: FP :)

File: server/sonar-server/src/main/java/org/sonar/server/component/ComponentFinder.java [hunk 2]
```diff
@@ -48,11 +49,22 @@ public ComponentDto getByUuidOrKey(DbSession dbSession, @Nullable String compone
...
+  }
+
+  public ComponentDto getByIdOrKey(DbSession dbSession, @Nullable Long componentId, @Nullable String componentKey, ParamNames parameterNames) {
+    checkArgument(componentId != null ^ componentKey != null, MSG_COMPONENT_ID_OR_KEY_TEMPLATE, parameterNames.getUuidParam(), parameterNames.getKeyParam());
+
+    if (componentId != null) {
+      return getById(dbSession, componentId);
+    }
+
+    checkArgument(!componentKey.isEmpty(), MSG_PARAMETER_MUST_NOT_BE_EMPTY, parameterNames.getKeyParam());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ComponentFinder.java]: MAJOR NullPointerException might be thrown as 'componentKey' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SelectAction.java [hunk 3]
```diff
@@ -19,42 +19,110 @@
...
+      .setProjectId(request.paramAsLong(PARAM_PROJECT_ID))
+      .setProjectKey(request.param(PARAM_PROJECT_KEY));
+  }
+
+  private ComponentDto getProject(DbSession dbSession, @Nullable Long projectId, @Nullable String projectKey) {
+    ComponentDto project = componentFinder.getByIdOrKey(dbSession, projectId, projectKey, ParamNames.PROJECT_ID_AND_KEY);
+
+    if (!userSession.hasPermission(GlobalPermissions.QUALITY_GATE_ADMIN) &&
+      !userSession.hasComponentUuidPermission(UserRole.ADMIN, project.uuid())) {
+      throw new ForbiddenException("Insufficient privileges");
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SelectAction.java]: `throw insufficientPrivilegesException()`

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SelectAction.java [hunk 4]
```diff
@@ -19,42 +19,110 @@
...
+    }
+
+    return project;
+  }
+
+  private static void checkQualityGate(DbClient dbClient, long id) {
+    QualityGateDto gate = dbClient.qualityGateDao().selectById(id);
+    if (gate == null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SelectAction.java]:
```
checkFound(dbClient.qualityGateDao().selectById(id), "There is no quality gate with id=" + id)
```

File: server/sonar-server/src/test/java/org/sonar/server/qualitygate/ws/SelectActionTest.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    ComponentDto project = insertProject();
+    QualityGateDto gate = insertQualityGate();
+    String gateId = String.valueOf(gate.getId());
+
+    userSession.login("login").addProjectUuidPermissions(UserRole.ADMIN, project.uuid());
+
+    callByKey(gateId, project.getKey());
+    assertSelected(gateId, project.getId());
+  }
+
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualitygate/ws/SelectActionTest.java]: add test as global admin

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SelectAction.java [hunk 6]
```diff
@@ -19,42 +19,110 @@
...
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+    this.componentFinder = componentFinder;
   }
 
   @Override
   public void define(WebService.NewController controller) {
     WebService.NewAction action = controller.createAction("select")
-      .setDescription("Associate a project to a quality gate. Require Administer Quality Gates permission")
+      .setDescription("Associate a project to a quality gate. Require Administer Quality Gates permission.")
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SelectAction.java]: specify that either project id or project key must be provided

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SelectAction.java]: remark out of scope of PR: it's misleading to have project id that are uuid and project id that are project db id depending of the places

**Comment by teryk** [comment]
apply feedbacks and LGTM

**Pull request merged**